crates/evm/src/precompile_cache.rs

```diff
@@ -0,0 +1,343 @@
+                                    upper_gas_limit: u64::MAX,
+                                    lower_gas_limit: gas_limit,
+                                },
+                            );
+                        }
+                    } else if result.result == InstructionResult::Return {
+                        // success
+                        if let Some(mut entry) = cache.cache.get(&key) {
+                            entry.result = Ok(result.clone());
+                            entry.upper_gas_limit = entry.upper_gas_limit.min(gas_limit);
+                        } else {
+                            cache.cache.insert(
+                                key,
+                                CacheEntry {
+                                    result: Ok(result.clone()),
+                                    upper_gas_limit: gas_limit,
+                                    lower_gas_limit: 0,
+                                },
+                            );
+                        }
+                    } else {
+                        // for other errors cache the result
+                        cache.cache.insert(
+                            key,
+                            CacheEntry {
+                                result: Ok(result.clone()),
+                                upper_gas_limit: gas_limit,
+                                lower_gas_limit: 0,
+                            },
+                        );
+                    }
+
+                    #[cfg(feature = "metrics")]
+                    {
+                        let new_entry_count = cache.cache.entry_count();
+                        if new_entry_count > entry_count_before {
+                            self.metrics
+                                .precompile_cache_size
+                                .increment((new_entry_count - entry_count_before) as f64);
+                        }
+                    }
+                }
+                Err(err) => {
+                    // fatal error
+                    #[cfg(feature = "metrics")]
+                    let entry_count_before = cache.cache.entry_count();
+
+                    cache.cache.insert(
+                        key,
+                        CacheEntry {
+                            result: Err(err.clone()),
+                            upper_gas_limit: gas_limit,
+                            lower_gas_limit: 0,
+                        },
+                    );
+
+                    #[cfg(feature = "metrics")]
+                    {
+                        let new_entry_count = cache.cache.entry_count();
+                        if new_entry_count > entry_count_before {
+                            self.metrics
+                                .precompile_cache_size
+                                .increment((new_entry_count - entry_count_before) as f64);
+                        }
+                    }
+                }
+                Ok(None) => {
```

Comment:
when does it happen? Should we still cache the `None` result?

Reply:
according to https://github.com/bluealloy/revm/blob/35a0611281997608dc9d1bd655b348105334f451/crates/handler/src/precompile_provider.rs#L99-L101 this happens when the precompile address is not found. i have added changes to cache this too

Reply:
turns out that we use the precompile provider with all the calls in a block, and the returned `Ok(None)` is the way to identify them as non-precompile calls, finally we won't cache these calls as the check in the inner provider is as cheap as the cache entry query